warehouse/oidc/views.py

```diff
@@ -68,81 +96,119 @@ def oidc_audience(request):
     require_csrf=False,
     has_translations=True,
 )
-def mint_token_from_oidc(request):
-    def _invalid(errors):
-        request.response.status = 422
-        return {"message": "Token request failed", "errors": errors}
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=True,
+)
+def mint_token_from_oidc(request: Request):
+    try:
+        payload = TokenPayload.model_validate_json(request.body)
+        unverified_jwt = payload.token
+    except ValidationError as exc:
+        return _invalid(
+            errors=[{"code": "invalid-payload", "description": str(exc)}],
+            request=request,
+        )
 
-    if request.flags.disallow_oidc(AdminFlagValue.DISALLOW_GITHUB_OIDC):
+    # We currently have an **unverified** JWT. To verify it, we need to
+    # know which OIDC service's keyring to check it against.
+    # To do this, we gingerly peek into the unverified claims and
+    # use the `iss` to key into the right `OIDCPublisherService`.
+    try:
+        unverified_claims = jwt.decode(
+            unverified_jwt, options=dict(verify_signature=False)
+        )
+        unverified_issuer: str = unverified_claims["iss"]
+    except Exception:
+        return _invalid(
+            errors=[{"code": "invalid-payload", "description": "malformed JWT"}],
+            request=request,
+        )
```

Comment:
NB: This is an inherently risky thing to do, but is also unfortunately the only way to get from a specific JWT to the service capable of verifying it.

This is safe *so long as* the `OIDCPublisherService` lookup that happens from this unverified claim never yields a service that no-ops the actual JWT signature and claim verification. That being said, this deserves multiple eyes on it to make sure I haven't missed something important here 🙂